Sync Gateway crashes while processing the changes feed

While running our app's test suite against a Sync Gateway build from c3599690e5, I consistently get the following crash:

```
2015-04-28T16:19:10.391-04:00 Changes: MultiChangesFeed done   (to 0cb04886-757c-43fe-84e3-c2eb1d034d9d)
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x88 pc=0x64503e]

goroutine 105 [running]:
github.com/couchbase/sync_gateway/db.func┬╖014()
        c:/BuildAgent/work/12217e9d79624586/src/github.com/couchbase/sync_gateway/db/changes.go:399 +0x173e
created by github.com/couchbase/sync_gateway/db.(*Database).MultiChangesFeed
        c:/BuildAgent/work/12217e9d79624586/src/github.com/couchbase/sync_gateway/db/changes.go:417 +0x5d6
```

The full log is available at https://gist.github.com/elevine/be82564f5022139094a4

--- a/src/github.com/couchbase/sync_gateway/db/changes.go
+++ b/src/github.com/couchbase/sync_gateway/db/changes.go
@@ -11,6 +11,7 @@ package db
 
 import (
 	"encoding/json"
+	"errors"
 	"fmt"
 	"sort"
 
@@ -40,6 +41,7 @@ type ChangeEntry struct {
 	Removed  base.Set    `json:"removed,omitempty"`
 	Doc      Body        `json:"doc,omitempty"`
 	Changes  []ChangeRev `json:"changes"`
+	Err      error       `json:"err,omitempty"` // Used to notify feed consumer of errors
 	branched bool
 }
 
@@ -141,6 +143,14 @@ func makeChangeEntry(logEntry *LogEntry, seqID SequenceID, channelName string) C
 	}
 	if logEntry.Flags&channels.Removed != 0 {
 		change.Removed = channels.SetOf(channelName)
+	}
+	return change
+}
+
+func makeErrorEntry(message string) ChangeEntry {
+
+	change := ChangeEntry{
+		Err: errors.New(message),
 	}
 	return change
 }
@@ -396,10 +406,16 @@ func (db *Database) MultiChangesFeed(chans base.Set, options ChangesOptions) (<-
 			// Before checking again, update the User object in case its channel access has
 			// changed while waiting:
 			if newCount := changeWaiter.CurrentUserCount(); newCount > userChangeCount {
-				base.LogTo("Changes+", "MultiChangesFeed reloading user %q", db.user.Name())
+				base.LogTo("Changes+", "MultiChangesFeed reloading user %+v", db.user)
 				userChangeCount = newCount
+				isAdmin := db.user == nil
 				if err := db.ReloadUser(); err != nil {
 					base.Warn("Error reloading user %q: %v", db.user.Name(), err)
+					if !isAdmin {
+						change := makeErrorEntry("User not found during reload - terminating changes feed")
+						base.LogTo("Changes+", "User not found during reload - terminating changes feed with entry %+v", change)
+						output <- &change
+					}
 					return
 				}
 			}

--- a/src/github.com/couchbase/sync_gateway/db/changes.go
+++ b/src/github.com/couchbase/sync_gateway/db/changes.go
@@ -41,7 +41,7 @@ type ChangeEntry struct {
 	Removed  base.Set    `json:"removed,omitempty"`
 	Doc      Body        `json:"doc,omitempty"`
 	Changes  []ChangeRev `json:"changes"`
-	Err      error       `json:"err,omitempty"` // Used to notify feed consumer of errors
+	Err      error       // Used to notify feed consumer of errors
 	branched bool
 }
 

--- a/src/github.com/couchbase/sync_gateway/db/database.go
+++ b/src/github.com/couchbase/sync_gateway/db/database.go
@@ -11,6 +11,7 @@ package db
 
 import (
 	"encoding/json"
+	"errors"
 	"expvar"
 	"fmt"
 	"net/http"
@@ -157,10 +158,15 @@ func (db *Database) ReloadUser() error {
 		return nil
 	}
 	user, err := db.Authenticator().GetUser(db.user.Name())
-	if err == nil {
-		db.user = user
+	if err != nil {
+		return err
 	}
-	return err
+	if user == nil {
+		return errors.New("User not found during reload")
+	} else {
+		db.user = user
+		return nil
+	}
 }
 
 //////// ALL DOCUMENTS:

--- a/src/github.com/couchbase/sync_gateway/rest/admin_api_test.go
+++ b/src/github.com/couchbase/sync_gateway/rest/admin_api_test.go
@@ -103,7 +103,7 @@ func TestUserDeleteDuringChangesWithAccess(t *testing.T) {
 
 	rt := restTester{syncFn: `function(doc) {channel(doc.channel); if(doc.type == "setaccess") { access(doc.owner, doc.channel);}}`}
 
-	response := rt.sendAdminRequest("PUT", "/_logging", `{"Changes+":true, "Changes":true, "HTTP":true}`)
+	response := rt.sendAdminRequest("PUT", "/_logging", `{"Changes+":true, "Changes":true, "Cache":true, "HTTP":true}`)
 
 	response = rt.sendAdminRequest("PUT", "/db/_user/bernard", `{"name":"bernard", "password":"letmein", "admin_channels":["foo"]}`)
 	assertStatus(t, response, 201)

--- a/src/github.com/couchbase/sync_gateway/rest/admin_api_test.go
+++ b/src/github.com/couchbase/sync_gateway/rest/admin_api_test.go
@@ -130,6 +130,8 @@ func TestUserDeleteDuringChangesWithAccess(t *testing.T) {
 		}
 	}()
 
+	// TODO: sleep required to ensure the changes feed iteration starts before the delete gets processed.
+	time.Sleep(100 * time.Millisecond)
 	rt.sendAdminRequest("PUT", "/db/bernard_doc1", `{"type":"setaccess", "owner":"bernard","channel":"foo"}`)
 	rt.sendAdminRequest("DELETE", "/db/_user/bernard", "")
 	rt.sendAdminRequest("PUT", "/db/manny_doc1", `{"type":"setaccess", "owner":"manny","channel":"bar"}`)

--- a/src/github.com/couchbase/sync_gateway/rest/admin_api_test.go
+++ b/src/github.com/couchbase/sync_gateway/rest/admin_api_test.go
@@ -10,8 +10,11 @@
 package rest
 
 import (
+	"bufio"
+	"bytes"
 	"encoding/json"
 	"fmt"
+	"io"
 	"log"
 	"testing"
 	"time"
@@ -93,6 +96,104 @@ func TestUserAPI(t *testing.T) {
 
 	// DELETE the user
 	assertStatus(t, rt.sendAdminRequest("DELETE", "/db/_user/snej", ""), 200)
+}
+
+// Test user delete while that user has an active changes feed (see issue 809)
+func TestUserDeleteDuringChangesWithAccess(t *testing.T) {
+
+	rt := restTester{syncFn: `function(doc) {channel(doc.channel); if(doc.type == "setaccess") { access(doc.owner, doc.channel);}}`}
+
+	response := rt.sendAdminRequest("PUT", "/_logging", `{"Changes+":true, "Changes":true, "HTTP":true}`)
+
+	response = rt.sendAdminRequest("PUT", "/db/_user/bernard", `{"name":"bernard", "password":"letmein", "admin_channels":["foo"]}`)
+	assertStatus(t, response, 201)
+
+	changesClosed := false
+	go func() {
+		response = rt.send(requestByUser("GET", "/db/_changes?feed=continuous&since=0", "", "bernard"))
+		// When testing single threaded, this reproduces the issue described in #809.
+		// When testing multithreaded (-cpu 4 -race), there are three (valid) possibilities"
+		// 1. The DELETE gets processed before the _changes auth completes: this will return 401
+		// 2. The _changes request gets processed before the DELETE: the changes response will be closed when the user is deleted
+		// 3. The DELETE is processed after the _changes auth completes, but before the MultiChangesFeed is instantiated.  The
+		//  changes feed doesn't have a trigger to attempt a reload of the user in this scenario, so will continue until disconnected
+		//  by the client.  This should be fixed more generally (to terminate all active user sessions when the user is deleted, not just
+		//  changes feeds) but that enhancement is too high risk to introduce at this time.
+		changesClosed = true
+		if response.Code == 401 {
+			// case 1 - ok
+		} else {
+			// case 2 - ensure no error processing the changes response.  Ensure no more than 2 entries
+			changes, err := readContinuousChanges(response)
+			assert.True(t, len(changes) <= 2)
+			assert.Equals(t, err, nil)
+		}
+	}()
+
+	rt.sendAdminRequest("PUT", "/db/bernard_doc1", `{"type":"setaccess", "owner":"bernard","channel":"foo"}`)
+	rt.sendAdminRequest("DELETE", "/db/_user/bernard", "")
+	rt.sendAdminRequest("PUT", "/db/manny_doc1", `{"type":"setaccess", "owner":"manny","channel":"bar"}`)
+	rt.sendAdminRequest("PUT", "/db/bernard_doc2", `{"type":"general", "channel":"foo"}`)
+
+	// case 3
+	for i := 0; i <= 5; i++ {
+		docId := fmt.Sprintf("/db/bernard_doc%d", i+3)
+		response = rt.sendAdminRequest("PUT", docId, `{"type":"setaccess", "owner":"bernard", "channel":"foo"}`)
+	}
+
+}
+
+// Reads continuous changes feed response into slice of ChangeEntry
+func readContinuousChanges(response *testResponse) ([]db.ChangeEntry, error) {
+	var change db.ChangeEntry
+	changes := make([]db.ChangeEntry, 0)
+	reader := bufio.NewReader(response.Body)
+	for {
+		entry, readError := reader.ReadBytes('\n')
+		if readError == io.EOF {
+			// done
+			break
+		}
+		if readError != nil {
+			// unexpected read error
+			return changes, readError
+		}
+		entry = bytes.TrimSpace(entry)
+		if len(entry) > 0 {
+			err := json.Unmarshal(entry, &change)
+			if err != nil {
+				return changes, err
+			}
+			changes = append(changes, change)
+		}
+
+	}
+	return changes, nil
+}
+
+// Test user deletion while that user has an active changes feed (see issue 809)
+func TestConcurrentUserDeleteDuringChanges(t *testing.T) {
+
+	rt := restTester{syncFn: `function(doc) {if(doc.type == "setaccess") {channel(doc.channel); access(doc.owner, doc.channel);} }`}
+
+	response := rt.sendAdminRequest("PUT", "/db/_user/bernard", `{"name":"bernard", "password":"letmein", "admin_channels":["foo"]}`)
+	assertStatus(t, response, 201)
+
+	response = rt.sendAdminRequest("PUT", "/_logging", `{"Changes+":true, "Changes":true, "HTTP":true}`)
+
+	log.Printf("logging response: %+v", response.Code)
+
+	go func() {
+		response = rt.send(requestByUser("GET", "/db/_changes?feed=continuous&since=0", "", "bernard"))
+		log.Printf("//////// _changes for bernard looks like: %s", response.Body.Bytes())
+	}()
+	time.Sleep(1 * time.Second)
+	rt.sendAdminRequest("PUT", "/db/doc1", `{"type":"setaccess", "owner":"bernard","channel":"foo"}`)
+
+	response = rt.sendAdminRequest("DELETE", "/db/_user/bernard", "")
+	rt.sendAdminRequest("PUT", "/db/doc2", `{"type":"setaccess", "owner":"manny","channel":"bar"}`)
+	time.Sleep(1 * time.Second)
+
 }
 
 func TestRoleAPI(t *testing.T) {

--- a/src/github.com/couchbase/sync_gateway/rest/admin_api_test.go
+++ b/src/github.com/couchbase/sync_gateway/rest/admin_api_test.go
@@ -171,31 +171,6 @@ func readContinuousChanges(response *testResponse) ([]db.ChangeEntry, error) {
 	return changes, nil
 }
 
-// Test user deletion while that user has an active changes feed (see issue 809)
-func TestConcurrentUserDeleteDuringChanges(t *testing.T) {
-
-	rt := restTester{syncFn: `function(doc) {if(doc.type == "setaccess") {channel(doc.channel); access(doc.owner, doc.channel);} }`}
-
-	response := rt.sendAdminRequest("PUT", "/db/_user/bernard", `{"name":"bernard", "password":"letmein", "admin_channels":["foo"]}`)
-	assertStatus(t, response, 201)
-
-	response = rt.sendAdminRequest("PUT", "/_logging", `{"Changes+":true, "Changes":true, "HTTP":true}`)
-
-	log.Printf("logging response: %+v", response.Code)
-
-	go func() {
-		response = rt.send(requestByUser("GET", "/db/_changes?feed=continuous&since=0", "", "bernard"))
-		log.Printf("//////// _changes for bernard looks like: %s", response.Body.Bytes())
-	}()
-	time.Sleep(1 * time.Second)
-	rt.sendAdminRequest("PUT", "/db/doc1", `{"type":"setaccess", "owner":"bernard","channel":"foo"}`)
-
-	response = rt.sendAdminRequest("DELETE", "/db/_user/bernard", "")
-	rt.sendAdminRequest("PUT", "/db/doc2", `{"type":"setaccess", "owner":"manny","channel":"bar"}`)
-	time.Sleep(1 * time.Second)
-
-}
-
 func TestRoleAPI(t *testing.T) {
 	var rt restTester
 	// PUT a role

--- a/src/github.com/couchbase/sync_gateway/rest/changes_api.go
+++ b/src/github.com/couchbase/sync_gateway/rest/changes_api.go
@@ -183,6 +183,9 @@ func (h *handler) sendSimpleChanges(channels base.Set, options db.ChangesOptions
 					break loop // end of feed
 				}
 				if nil != entry {
+					if entry.Err != nil {
+						break loop // error returned by feed - end changes
+					}
 					if first {
 						first = false
 					} else {
@@ -269,6 +272,8 @@ loop:
 				feed = nil
 			} else if entry == nil {
 				err = send(nil)
+			} else if entry.Err != nil {
+				break loop // error returned by feed - end changes
 			} else {
 				entries := []*db.ChangeEntry{entry}
 				waiting := false
@@ -283,6 +288,8 @@ loop:
 						} else if entry == nil {
 							waiting = true
 							break collect
+						} else if entry.Err != nil {
+							break loop // error returned by feed - end changes
 						}
 						entries = append(entries, entry)
 					default:

